### PR TITLE
ntf: update subscription status, error handling on SMP errors / agent events

### PR DESCRIPTION
**Pull request by jr-simplex**

### CODE REVIEW

File: src/Simplex/Messaging/Notifications/Server.hs [hunk 1]
```diff
@@ -109,43 +109,58 @@ ntfSubscriber NtfSubscriber {smpSubscribers, newSubQ, smpAgent = ca@SMPClientAge
...
-            pure ()
+          CAResubscribed srv sub -> do
+            let ntfId = snd sub
+                smpQueue = SMPQueueNtf srv ntfId
+            updateSubStatus smpQueue NSActive
+          CASubError srv sub err -> do
+            logError (T.pack $ "ntfSubscriber receiveAgent CASubError: " <> show err)
+            let ntfId = snd sub
+                smpQueue = SMPQueueNtf srv ntfId
+            updateSubStatus smpQueue NSInactive
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Notifications/Server.hs:156]: should we add another status?

File: src/Simplex/Messaging/Agent/NtfSubSupervisor.hs [hunk 2]
```diff
@@ -168,8 +168,10 @@ runNtfWorker c srv doWork = forever $ do
                       NSNew -> updateSubNextCheck ts NSNew
                       NSPending -> updateSubNextCheck ts NSPending
                       NSActive -> updateSubNextCheck ts NSActive
+                      NSInactive -> updateSubNextCheck ts NSInactive
                       NSEnd -> updateSubNextCheck ts NSEnd
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/NtfSubSupervisor.hs:172]: I think it's overkill tbh

File: src/Simplex/Messaging/Agent/NtfSubSupervisor.hs [hunk 3]
```diff
@@ -168,8 +168,10 @@ runNtfWorker c srv doWork = forever $ do
...
                       NSSMPAuth -> updateSub (NASCreated NSSMPAuth) (NtfSubAction NSADelete) ts -- TODO re-create subscription?
+                      NSSMPErr e -> updateSubNextCheck ts $ NSSMPErr e
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/NtfSubSupervisor.hs:174]: all these errors are permanent I think? version check should be made separately

File: src/Simplex/Messaging/Notifications/Server.hs [hunk 4]
```diff
@@ -109,43 +110,71 @@ ntfSubscriber NtfSubscriber {smpSubscribers, newSubQ, smpAgent = ca@SMPClientAge
...
           ntfTs <- liftIO getSystemTime
-          let smpQueue = SMPQueueNtf srv ntfId
           st <- asks store
           NtfPushServer {pushQ} <- asks pushServer
           atomically $
             findNtfSubscriptionToken st smpQueue
               >>= mapM_ (\tkn -> writeTBQueue pushQ (tkn, PNMessage PNMessageData {smpQueue, ntfTs, nmsgNonce, encNMsgMeta}))
+        SMP.END -> updateSubStatus smpQueue NSInactive
+        SMP.ERR AUTH -> updateSubStatus smpQueue NSSMPAuth
+        SMP.ERR e -> updateSubStatus smpQueue (NSSMPErr e)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Notifications/Server.hs:136]: these are not possible - errors are not sent as messages

Comment (epoberezkin) [inline src/Simplex/Messaging/Notifications/Server.hs:136]:
```suggestion
```

File: src/Simplex/Messaging/Notifications/Server.hs [hunk 5]
```diff
@@ -109,43 +110,71 @@ ntfSubscriber NtfSubscriber {smpSubscribers, newSubQ, smpAgent = ca@SMPClientAge
...
+                smpQueue = SMPQueueNtf srv ntfId
+            updateSubStatus smpQueue NSActive
+          CASubError srv sub err -> do
+            let ntfId = snd sub
+                smpQueue = SMPQueueNtf srv ntfId
+            case err of
+              PCEProtocolError e -> case e of
+                AUTH -> updateSubStatus smpQueue NSSMPAuth
+                _ -> updateSubStatus smpQueue (NSSMPErr e)
+              PCEResponseError e -> updateSubStatus smpQueue (NSSMPErr e)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Notifications/Server.hs:160]:
```suggestion
              PCEResponseError e -> logErr err
```

Comment (epoberezkin) [reply src/Simplex/Messaging/Notifications/Server.hs:160]: the reason is that we don't know what is the problem here, it's some response we couldn't parse - it's not the same as protocol error

File: src/Simplex/Messaging/Notifications/Server.hs [hunk 6]
```diff
@@ -109,43 +110,71 @@ ntfSubscriber NtfSubscriber {smpSubscribers, newSubQ, smpAgent = ca@SMPClientAge
...
+              PCEUnexpectedResponse -> logErr err >> updateSubStatus smpQueue NSInactive
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Notifications/Server.hs:161]:
```suggestion
              PCEUnexpectedResponse -> logErr err
```

File: src/Simplex/Messaging/Notifications/Server.hs [hunk 7]
```diff
@@ -109,43 +110,69 @@ ntfSubscriber NtfSubscriber {smpSubscribers, newSubQ, smpAgent = ca@SMPClientAge
...
+              PCEResponseTimeout -> updateSubStatus smpQueue NSInactive
+              PCENetworkError -> updateSubStatus smpQueue NSInactive
+              PCETransportError e -> logErr e >> updateSubStatus smpQueue NSInactive
+              PCESignatureError e -> logErr e >> updateSubStatus smpQueue NSInactive
+              PCEIOError e -> logErr e >> updateSubStatus smpQueue NSInactive
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Notifications/Server.hs:164]:
```suggestion
              PCESignatureError e -> logErr e
              PCEIOError e -> logErr e
              _ -> pure ()
```

**Pull request merged**